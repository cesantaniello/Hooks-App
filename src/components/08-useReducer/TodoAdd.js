import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {
    
    const [{desc}, handleInputChange, reset] = useForm({
        desc: ''
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (desc.trim().length <= 1) {
            return;
        } 

        const newTodo = {
            id: new Date().getTime(),
            desc: desc,
            done: false
        }

        handleAddTodo(newTodo);
        reset();        
    }

    return (
        <>
            <h4>Agregar ToDo</h4>
            <hr/>

            <form className="d-grid gap-2" onSubmit={handleSubmit}>
                <input
                    type="text" 
                    name="desc" 
                    placeholder="Aprender..." 
                    autocomplete="off"
                    className="form-control"
                    value={desc}
                    onChange={handleInputChange}
                />
                <button type="submit" className="btn btn-outline-primary mt-1">
                    Agregar
                </button>
            </form>            
        </>
    )
}
