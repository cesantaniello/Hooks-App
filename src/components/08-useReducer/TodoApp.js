import React, {useReducer, useEffect} from 'react';
import {todoReducer} from './todoReducer';
import {useForm} from '../../hooks/useForm';
import { TodoList } from './TodoList';

import './styles.css';

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
}

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    const [{desc}, handleInputChange, reset] = useForm({
        desc: ''
    });

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (todoId) => {
        dispatch({
            type: 'delete',
            payload: todoId
        });
    }

    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        });
    }

    const handleSubmit = (e) => {
        e.preventDefault();

        if (desc.trim().length <= 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: desc,
            done: false
        }

        const action = {
            type: 'add',
            payload: newTodo
        }

        dispatch(action);
        reset();
    }

    return (
        <div>
            <h1>TodoApp ({todos.length})</h1>
            <hr/>
            <div className="row">
                <div className="col-7">
                    <TodoList 
                        todos={todos}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}
                    />
                </div>
                <div className="col-5">
                    <h4>Agregar ToDo</h4>
                    <hr/>

                    <form className="d-grid gap-2" onSubmit={handleSubmit}>
                        <input
                            type="text" 
                            name="desc" 
                            placeholder="Aprender..." 
                            autocomplete="off"
                            className="form-control"
                            value={desc}
                            onChange={handleInputChange}
                        />
                        <button type="submit" className="btn btn-outline-primary mt-1">
                            Agregar
                        </button>
                    </form>
                </div>
            </div>
        </div>
    )
}
