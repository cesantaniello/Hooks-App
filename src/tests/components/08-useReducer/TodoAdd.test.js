import React from 'react';
import {shallow} from 'enzyme';
import {TodoAdd} from '../../../components/08-useReducer/TodoAdd';


describe('Tests on <TodoAdd />', () => {
    const handleAddTodo = jest.fn();

    const wrapper = shallow(
        <TodoAdd 
            handleAddTodo={handleAddTodo} 
        />
    );

    test('Should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('Should not call handleAddTodo on submit', () => {
        wrapper.find('form').simulate('submit', {
            preventDefault: () => {}
        });

        expect(handleAddTodo).toHaveBeenCalledTimes(0);
    })

    test('Should call handleAddTodo on submit', () => {
        wrapper.find('input').simulate('change', {
            target: {
                name: 'desc',
                value: 'Aprender React'
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault: () => {}
        });

        expect(handleAddTodo).toHaveBeenCalledTimes(1);
        expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
        expect(handleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: 'Aprender React',
            done: false
        })

        expect(wrapper.find('input').prop('value')).toBe('');
    })
})