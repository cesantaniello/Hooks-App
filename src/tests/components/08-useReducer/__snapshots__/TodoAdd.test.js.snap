// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <TodoAdd /> Should render correctly 1`] = `
<Fragment>
  <h4>
    Agregar ToDo
  </h4>
  <hr />
  <form
    className="d-grid gap-2"
    onSubmit={[Function]}
  >
    <input
      autocomplete="off"
      className="form-control"
      name="desc"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
